server/src/main/java/org/elasticsearch/index/engine/LuceneSyntheticSourceChangesSnapshot.java

```diff
@@ -0,0 +1,249 @@
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Comparator;
+import java.util.Deque;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Set;
+
+/**
+ * A {@link SearchBasedChangesSnapshot} that utilizes a synthetic field loader to rebuild the recovery source.
+ * This snapshot is activated when {@link IndexSettings#RECOVERY_SOURCE_SYNTHETIC_ENABLED_SETTING}
+ * is enabled on the underlying index.
+ *
+ * The {@code maxMemorySizeInBytes} parameter limits the total size of uncompressed _sources
+ * loaded into memory during batch retrieval.
+ */
+public class LuceneSyntheticSourceChangesSnapshot extends SearchBasedChangesSnapshot {
+    public static final int DEFAULT_SEARCH_BATCH_SIZE = 1024;
+    public static final long DEFAULT_MEMORY_SIZE = 4 * 1024 * 1024; // 4MB
+
+    private final long maxMemorySizeInBytes;
+    private final StoredFieldLoader storedFieldLoader;
+    private final SourceLoader sourceLoader;
+
+    private int skippedOperations;
+    private long lastSeenSeqNo;
+
+    private record SearchRecord(FieldDoc doc, boolean isTombstone, long seqNo, long primaryTerm, long version, long size) {
+        int index() {
+            return doc.shardIndex;
+        }
+
+        int docID() {
+            return doc.doc;
+        }
+
+        boolean hasRecoverySourceSize() {
+            return size != -1;
+        }
+    }
+
+    private final Deque<SearchRecord> pendingDocs = new LinkedList<>();
+    private final Deque<Translog.Operation> operationQueue = new LinkedList<>();
+
+    public LuceneSyntheticSourceChangesSnapshot(
+        MappingLookup mappingLookup,
+        Engine.Searcher engineSearcher,
+        int searchBatchSize,
+        long maxMemorySizeInBytes,
+        long fromSeqNo,
+        long toSeqNo,
+        boolean requiredFullRange,
+        boolean accessStats,
+        IndexVersion indexVersionCreated
+    ) throws IOException {
+        super(engineSearcher, searchBatchSize, fromSeqNo, toSeqNo, requiredFullRange, accessStats, indexVersionCreated);
+        assert mappingLookup.isSourceSynthetic();
+        this.maxMemorySizeInBytes = maxMemorySizeInBytes;
+        this.sourceLoader = mappingLookup.newSourceLoader(SourceFieldMetrics.NOOP);
+        Set<String> storedFields = sourceLoader.requiredStoredFields();
+        this.storedFieldLoader = StoredFieldLoader.create(mappingLookup.isSourceSynthetic() == false, storedFields);
+        this.lastSeenSeqNo = fromSeqNo - 1;
+    }
+
+    @Override
+    public int skippedOperations() {
+        return skippedOperations;
+    }
+
+    @Override
+    public Translog.Operation nextOperation() throws IOException {
+        while (true) {
+            if (operationQueue.isEmpty()) {
+                loadNextBatch();
+            }
+            if (operationQueue.isEmpty()) {
+                return null;
+            }
+            var op = operationQueue.pollFirst();
```

Comment:
Sorry for my ignorance here, but `InternalEngine#newChangesSnapshot` has a `singleConsumer` parameter that indicates if the snapshot will be consumed by multiple threads. In such case, I wonder if we should concurrent queues here?

Reply:
Yes, there is confusion between singleConsumer and single-threaded access. A snapshot of changes should be accessed by at most one thread at a time, but the accessing thread can change during peer recovery. We introduced the `singleConsumer` flag to prevent reusing sequentially stored fields created by other threads in this case. I think it's fine to use a non-concurrent queue.

Reply:
Thanks @dnhatn ! 

If we expect `SearchBasedChangesSnapshot#next` to be called by a single thread at a time, then would it make sense to add documentation for this and have some assertions around concurrent accesses?

Reply:
I can add documentation. Regarding assertions, do you mean in the code to check that we don't detect concurrent access? Is it even possible?

Reply:
We have `assert assertAccessingThread();` for this purpose.

Reply:
Yep I can add that, my comment was more that we cannot check if the class is accessed concurrently. That's more a contract that we need to document?